File: src/lib/pipeline/orchestrator.ts
// ============================================================
// Pipeline Orchestrator
// Runs the 5-stage pipeline sequentially for a given request ID
// ============================================================

import { updateIntakeRequest } from '@/lib/db/intake';
import { createPipelineRun, updatePipelineRun } from '@/lib/db/pipeline-runs';
import { runStage1 } from './stage1-intelligence';
import { runStage2 } from './stage2-images';
import { runStage3 } from './stage3-copy';
import { runStage4 } from './stage4-compose';
import { runStage5 } from './stage5-surface';

interface PipelineResult {
  success: boolean;
  failed_stage?: number;
  error?: string;
}

type StageFn = (requestId: string) => Promise<Record<string, unknown>>;

interface StageDefinition {
  num: number;
  name: string;
  fn: StageFn;
}

export async function runPipeline(requestId: string): Promise<PipelineResult> {
  // Mark the request as generating
  await updateIntakeRequest(requestId, { status: 'generating' });

  const stages: StageDefinition[] = [
    { num: 1, name: 'Strategic Intelligence', fn: runStage1 },
    { num: 2, name: 'Character-Driven Image Generation', fn: runStage2 },
    { num: 3, name: 'Copy Generation', fn: runStage3 },
    { num: 4, name: 'Layout Composition', fn: runStage4 },
    { num: 5, name: 'Surface & Distribute', fn: runStage5 },
  ];

  for (const stage of stages) {
    const startTime = Date.now();

    // Create a pipeline run record (status: running)
    const run = await createPipelineRun({
      request_id: requestId,
      stage: stage.num,
      stage_name: stage.name,
      status: 'running',
    });

    try {
      const output = await stage.fn(requestId);
      const durationMs = Date.now() - startTime;

      // Mark stage as passed
      await updatePipelineRun(run.id, {
        status: 'passed',
        output_data: output,
        duration_ms: durationMs,
        completed_at: new Date().toISOString(),
      });
    } catch (err) {
      const durationMs = Date.now() - startTime;
      const errorMessage = err instanceof Error ? err.message : String(err);

      // Mark stage as failed
      await updatePipelineRun(run.id, {
        status: 'failed',
        error_message: errorMessage,
        duration_ms: durationMs,
        completed_at: new Date().toISOString(),
      });

      // Mark the request as failed (back to draft so user can retry)
      await updateIntakeRequest(requestId, { status: 'draft' });

      return {
        success: false,
        failed_stage: stage.num,
        error: `Stage ${stage.num} (${stage.name}) failed: ${errorMessage}`,
      };
    }
  }

  // All stages passed — mark request as ready for review
  await updateIntakeRequest(requestId, { status: 'review' });

  return { success: true };
}

File: worker/pipeline/orchestrator.py
"""Pipeline orchestrator -- runs stages 1-6 sequentially.

Handles full generation and single-stage regeneration. Each stage
receives a shared *context* dict that accumulates results. On failure
the request is returned to 'draft' status and Teams is notified.
"""
from __future__ import annotations

import logging

from neon_client import get_actors, update_request_status
from teams_notify import notify_generation_complete, notify_generation_failed

from pipeline.stage1_intelligence import run_stage1
from pipeline.stage2_images import run_stage2
from pipeline.stage3_copy import run_stage3
from pipeline.stage4_compose_v3 import run_stage4
from pipeline.stage4_organic_carousel import run_organic_carousels
from pipeline.stage5_video import run_stage5 as run_video_stage
from pipeline.stage6_landing_pages import run_stage6

logger = logging.getLogger(__name__)


async def run_pipeline(job: dict) -> None:
    """Run the full Creative OS pipeline for a compute job.

    Parameters
    ----------
    job:
        A dict from ``fetch_pending_jobs`` containing at minimum
        ``request_id``, ``job_type``, and optionally ``stage_target``
        and ``feedback``.
    """
    request_id: str = job["request_id"]
    job_type: str = job["job_type"]
    stage_target = job.get("stage_target")
    feedback = job.get("feedback")

    # ── Country Job Creator: create per-country jobs on the SAME request ──
    # If country_quotas exist, create one compute_job per country.
    # Each country job runs the full pipeline independently.
    if job_type == "generate":
        from neon_client import get_intake_request as _get_request

        from pipeline.country_job_creator import create_country_jobs, has_country_quotas
        request = await _get_request(request_id)
        if has_country_quotas(request):
            logger.info("Multi-country campaign detected — creating per-country jobs")
            country_jobs = await create_country_jobs(request, request_id)
            logger.info(
                "Created %d country jobs. Parent job done. "
                "Country jobs will be picked up on next poll cycle.",
                len(country_jobs),
            )
            return

    stages = [
        (1, "Strategic Intelligence", run_stage1),
        (2, "Character-Driven Image Generation", run_stage2),
        (3, "Copy Generation", run_stage3),
        (4, "Layout Composition", run_stage4),
        (5, "Video Generation", run_video_stage),
        (6, "Landing Page Generation", run_stage6),
    ]

    # If regenerating a specific stage, only run that one.
    if job_type == "regenerate_stage" and stage_target is not None:
        target = int(stage_target)
        stages = [(n, name, fn) for n, name, fn in stages if n == target]

    # If resuming from a stage, run that stage through the end.
    if job_type == "resume_from" and stage_target is not None:
        target = int(stage_target)
        stages = [(n, name, fn) for n, name, fn in stages if n >= target]
        logger.info("Resuming pipeline from Stage %d (stages %s)", target, [n for n, _, _ in stages])

    context: dict = {
        "request_id": request_id,
        "feedback": feedback,
    }

    # ── Country Job: inject country context into pipeline ──
    if job_type == "generate_country":
        from neon_client import get_intake_request as _get_request
        request = await _get_request(request_id)
        country = job.get("country", "")
        feedback_data = job.get("feedback_data", {})
        if isinstance(feedback_data, str):
            import json as _json
            feedback_data = _json.loads(feedback_data)

        # Inject country + scaling into context so stages can read it
        context["country"] = country
        context["persona_count"] = feedback_data.get("persona_count", 2)
        context["actors_per_persona"] = feedback_data.get("actors_per_persona", 2)
        context["country_quota"] = feedback_data
        # Override target_regions to just this country
        context["target_regions"] = [country]
        context["form_data"] = request.get("form_data", {})
        logger.info("Running pipeline for country: %s (personas=%d, actors=%d)",
                     country, context["persona_count"], context["actors_per_persona"])

    # ── Load brief + request data from Neon ──────────────────────
    # Always load if starting at stage > 1 (resume, regenerate, or auto-skip)
    first_stage = stages[0][0] if stages else 1
    if first_stage >= 1:
        try:
            from neon_client import get_brief, get_intake_request
            existing_brief = await get_brief(request_id)
            if existing_brief:
                import json
                brief_data = existing_brief.get("brief_data")
                if isinstance(brief_data, str):
                    brief_data = json.loads(brief_data)
                if brief_data and "campaign_objective" in brief_data:
                    request = await get_intake_request(request_id)
                    context.update({
                        "request_title": request.get("title", "Untitled"),
                        "brief": brief_data,
                        "personas": brief_data.get("personas", []),
                        "cultural_research": brief_data.get("cultural_research", {}),
                        "design_direction": existing_brief.get("design_direction", {}),
                        "target_languages": request.get("target_languages", []),
                        "target_regions": request.get("target_regions", []),
                        "form_data": request.get("form_data", {}),
                    })
                    if first_stage == 1:
                        logger.info("SKIPPING Stage 1 — brief already exists in Neon (campaign: %s)", brief_data.get("campaign_objective", "")[:80])
                        stages = [(n, name, fn) for n, name, fn in stages if n > 1]
                    else:
                        logger.info("Loaded brief from Neon for resume/regen (campaign: %s)", brief_data.get("campaign_objective", "")[:80])
        except Exception as e:
            logger.info("Could not check for existing brief: %s — continuing with empty brief", e)

    for stage_num, stage_name, stage_fn in stages:
        logger.info("Running Stage %d: %s", stage_num, stage_name)

        # Stages 4+ need actors in context — load from Neon if not already present
        if stage_num >= 4 and "actors" not in context:
            try:
                actors = await get_actors(request_id)
                context["actors"] = actors
                logger.info("Loaded %d actors from Neon for stage %d", len(actors), stage_num)
            except Exception as e:
                logger.warning("Could not load actors: %s", e)
                context["actors"] = []

        try:
            result = await stage_fn(context)
            context.update(result)  # each stage can pass data to the next
            logger.info("Stage %d passed.", stage_num)

            # After Stage 4: generate organic carousels (LinkedIn + IG)
            if stage_num == 4:
                try:
                    logger.info("Generating organic carousels...")
                    organic_result = await run_organic_carousels(context)
                    context.update(organic_result)
                    logger.info(
                        "Organic carousels: %d generated",
                        organic_result.get("organic_carousel_count", 0),
                    )
                except Exception as org_exc:
                    logger.warning("Organic carousel generation failed (non-fatal): %s", org_exc)

        except Exception as exc:
            logger.error("Stage %d failed: %s", stage_num, exc, exc_info=True)
            await update_request_status(request_id, "draft")
            await notify_generation_failed(
                context.get("request_title", "Unknown"),
                str(exc),
            )
            raise

    # Status rollup: for country jobs, only move to 'review' when ALL countries are done
    if job_type == "generate_country":
        from neon_client import check_all_country_jobs_complete
        if await check_all_country_jobs_complete(request_id):
            await update_request_status(request_id, "review")
            await notify_generation_complete(
                context.get("request_title", "Unknown"),
                context.get("asset_count", 0),
                request_id,
            )
            logger.info("All country jobs complete — campaign moved to 'review'")
        else:
            logger.info("Country job for %s done, but other countries still processing",
                        job.get("country", "unknown"))
    else:
        await update_request_status(request_id, "review")
        await notify_generation_complete(
            context.get("request_title", "Unknown"),
            context.get("asset_count", 0),
            request_id,
        )
